templates/core/Resources/_ti/bootstrap.loader.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2018 by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ * 
+ * Description:
+ * This script recursively searches the "Resources" directory for all JavaScript files
+ * named "bootstrap.js" or ending with the name "*.bootstrap.js" and then executes them.
+ * The main intention of this feature is to all JavaScript to kick-off functionality or
+ * display UI to the end-user before the "app.js" gets loaded. This feature is the CommonJS
+ * equivalent to Titanium's Android module onAppCreate() or iOS module load() features.
+ * 
+ * Use-Cases:
+ * - Automatically kick-off analytics functionality on app startup.
+ * - Ensure "Google Play Services" is installed/updated on app startup on Android.
+ */
+
+'use strict';
+
+var JS_EXTENSION = '.js',
+	BOOTSTRAP_FILE_NAME = 'bootstrap' + JS_EXTENSION,
+	BOOTSTRAP_SUFFIX = '.' + BOOTSTRAP_FILE_NAME;
+
+exports.loadAsync = function (finished) {
+	var resourceDirectory = Ti.Filesystem.getFile(Ti.Filesystem.getResourcesDirectory()),
+		resourceDirectoryPath = resourceDirectory.nativePath,
+		bootstrapScripts = [];
+
+	// Fetches all "*.bootstrap.js" files under given directory and adds them to the "bootstrapScripts" array.
+	// The names stored in the array can be loaded via require() method.
+	function fetchBootstrapScriptsFrom(file) {
+		var index,
+			fileNameArray,
+			bootstrapPath;
+
+		if (file) {
+			if (file.isDirectory()) {
+				// This is a directory. Recursively look for bootstrap files under it.
+				fileNameArray = file.getDirectoryListing();
+				if (fileNameArray) {
+					for (index = 0; index < fileNameArray.length; index++) {
+						fetchBootstrapScriptsFrom(Ti.Filesystem.getFile(file.nativePath, fileNameArray[index]));
+					}
+				}
+			} else if ((file.name === BOOTSTRAP_FILE_NAME) || file.name.endsWith(BOOTSTRAP_SUFFIX)) {
+				// This is a bootstrap file.
+				// Convert its path to something loadable via require() and add it to the array.
+				bootstrapPath = file.nativePath;
+				bootstrapPath = bootstrapPath.substr(
+					resourceDirectoryPath.length,
+					(bootstrapPath.length - resourceDirectoryPath.length) - JS_EXTENSION.length);
+				bootstrapScripts.push(bootstrapPath);
+			}
+		}
+	}
+	fetchBootstrapScriptsFrom(resourceDirectory);
+
+	// Sort the bootstrap scripts so that they'll be loaded in a consistent order between platforms.
+	bootstrapScripts.sort();
```

Comment:
Maybe some docs/help on what exactly the order will be? Obviously different OSes may sort differently naturally, so this is a potential point of confusion.

Reply:
I'm curious about that too. I'm not quite sure if calling JavaScript _Array.sort()_ without any arguments guarantees the order across platforms actually. I see that [Mozilla describes it's sorted according to Unicode code point](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and Microsoft sorts it in [ASCII character order](https://docs.microsoft.com/en-us/scripting/javascript/reference/sort-method-array-javascript), and ECMAScript spec says it orders [according to _ToString()_ comparison](https://www.ecma-international.org/ecma-262/5.1/#sec-15.4.4.11) (Do they describe same thing? I'm not sure). Maybe we might want to place _compareFunction_ explicitly to avoid confusion.

Reply:
I was thinking that we want to give scripts a chance to define order of the bootstrap script by defining priority like `export.priority` but for now I'm almost fine with current implementation. I think simply sorting files according to file names would work well for developers. For instance in case developer wants to bootstrap network scripts before UI scripts, just rename bootstrap files like

```
001.something_to_boot_Network.js
002.something_to_boot_UI_theme.js
003.google_play.js
```